src/_hyperscript.js

```diff
@@ -1244,6 +1244,31 @@
                     }
                 });
 
+                _parser.addGrammarElement("arrayIndex", function (parser, tokens, root) {
+
+                    if (tokens.matchOpToken("[")) {
+                        var indexExp = parser.parseElement("expression", tokens)
+                        tokens.requireOpToken("]")
+
+                        var arrayIndex = {
+                            type: "arrayIndex",
+                            root: root,
+                            index: indexExp,
+                            evaluate: function (context) {
+                                var index = _runtime.unifiedEval(this, function(context, rootVal) {
+                                    return parseInt(rootVal)
+                                }, context, this.index)
+
+                                return _runtime.unifiedEval(this, function(context, rootVal){
+                                    return rootVal[index];
+                                }, context, root)
+                            }
```

Comment:
Close!  The trick is to pass both root and index in to a single `unifiedEval`:

```js
return _runtime.unifiedEval(this, function(context, rootVal, index){
                                    return rootVal[index];
                                }, context, root, indexExp)
```
It isn't documented at all, but the arguments that are passed in after the function are correctly resolved and then passed back to that function, in order.  So you get context, root and indexExp all evaluated async-transparently and handed back to you.

Also, I don't think we need to parse an int here, we can rely on the user to supply the correct number type.